curp/src/channel/key_mpsc.rs

```diff
@@ -0,0 +1,359 @@
+}
+
+hash_eq!(MpscKeysMessage);
+
+impl<K, M> Clone for MpscKeysMessage<K, M> {
+    fn clone(&self) -> Self {
+        Self {
+            inner: Arc::<_>::clone(&self.inner),
+            ready: Arc::<_>::clone(&self.ready),
+        }
+    }
+}
+
+#[allow(unsafe_code)]
+unsafe impl<K: Send, M: Send> Send for MpscKeysMessage<K, M> {}
+
+impl<K, M> MpscKeysMessage<K, M> {
+    /// Create a new `KeysMessage`
+    fn new(keys: Vec<K>, message: M) -> Self {
+        Self {
+            inner: Arc::new(KeysMessageInner::new(keys, message)),
+            ready: Arc::new(Mutex::new(false)),
+        }
+    }
+
+    /// modify data in the message if necessary
+    pub(crate) fn keys(&self) -> &[K] {
+        self.inner.keys()
+    }
+
+    /// map the inner message to a closure
+    pub(crate) fn map_msg<R, F: FnOnce(&mut M) -> R>(&self, f: F) -> R {
+        self.inner.map_msg(f)
+    }
+
+    /// Tell if the message is ready
+    pub(crate) fn ready(&self) -> bool {
+        *self.ready.lock()
+    }
+
+    /// Mark the message as ready
+    pub(crate) fn mark_ready(&self) {
+        *self.ready.lock() = true;
+    }
+}
+
+impl<K: Eq + Hash + Clone + ConflictCheck, M> KeybasedChannel<MpscKeysMessage<K, M>> {
+    /// Insert successors and proceeders info
+    fn insert_graph(&mut self, new_km: MpscKeysMessage<K, M>) {
+        let proceeder_cnt: u64 = self
+            .successor
+            .iter_mut()
+            .filter_map(|(k, v)| {
+                super::keys_conflict(k.keys(), new_km.keys()).then(|| {
+                    let _ignore = v.insert(new_km.clone());
+                    1
+                })
+            })
+            .sum();
+        // the message can only be inserted once, so we ignore the return value
+        let _ignore = self.successor.insert(new_km.clone(), HashSet::new());
+
+        if proceeder_cnt != 0 {
+            // the message can only be inserted once, so we ignore the return value
+            let _ignore2 = self.proceeder.insert(new_km, proceeder_cnt);
+        }
+    }
+
+    /// Append a key and message to this `KeybasedChannel`
+    fn append(&mut self, keys: &[K], msg: M) -> MpscKeysMessage<K, M> {
+        let km = MpscKeysMessage::new(keys.to_vec(), msg);
+        self.insert_graph(km.clone());
+        km
+    }
+
+    /// Move a message for the `key` from pending to inner recursively
+    fn mark_ready(&mut self, km: &MpscKeysMessage<K, M>) {
+        km.mark_ready();
+        let mut ready_pool = VecDeque::new();
+        let mut ready_cnt = 0;
+        if self.proceeder.get(km).is_none() {
+            ready_pool.push_back(km.clone());
+            ready_cnt = 1;
+        }
+
+        while let Some(cur) = ready_pool.pop_front() {
+            self.inner.push_back(cur.clone());
+            if let Some(successor) = self.successor.remove(km) {
+                for s in successor {
+                    let (no_proceeder, has_pre) = if let Some(s_p) = self.proceeder.get_mut(&s) {
```

Comment:
Is it possible that `s` in successor but not in proceeder?

Reply:
It's possible. If one node has successors, it's in successor structure. The same for preceeder.

Reply:
if a node in another node's successor list, it must have at least one proceeder right?

Reply:
Yeah, you're right, they're a pair. I check all the possible cases to avoid using `unwrap`.

Reply:
I'll replace it with a `unreachable` statement.